torch/csrc/jit/codegen/onednn/graph_helper.cpp

```diff
@@ -0,0 +1,521 @@
+  switch (type) {
+    case DeviceType::CPU:
+      return dnnl::graph::engine::kind::cpu;
+    default:
+      TORCH_CHECK(false, "Not support device type ", type);
+  }
+}
+
+void mayAddListConstructIntoConcatPartition(
+    Node* n,
+    OpPartitionMap& opToOwningPartition) {
+  // Since prim::ListConstruct is not visible to the LLGA,
+  // it will not be in any partition returned from partfuseritioning results.
+  // We need rewrite opToOwningPartition to make the prim::ListConstruct to be
+  // 'virtually' in the same partition with the aten::cat, so that
+  // prim::ListConstruct can be fused into the fusion group by graph fuser.
+  // We emphasize on 'virtually' because get_num_ops() for cat's partition
+  // would still return 1.
+  if (n->kind() == aten::cat && opToOwningPartition.has(n)) {
+    auto listConstrcut = n->input(0)->node();
+    auto partitionId = opToOwningPartition.get(n);
+    opToOwningPartition.add(listConstrcut, partitionId);
+  }
+}
+
+LlgaGraphHelper::LlgaGraphHelper(
+    const std::shared_ptr<Graph>& graph,
+    dnnl::graph::partition::policy policy) {
+  auto deviceType = inferDevice(graph);
+  auto engineKind = getLlgaEngineKind(deviceType);
+  dnnl::graph::graph g{engineKind};
+
+  GRAPH_DEBUG("Constructing LLGA graph");
+  // TODO: select nodes in top-level block for now
+  for (auto* node : graph->block()->nodes()) {
+    auto op = createLlgaOp(node);
+    try {
+      g.add_op(op);
+    } catch (std::exception& e) {
+      GRAPH_DEBUG(
+          "The backend failed to add node ", node->kind().toQualString());
+      g.add_op(makeWildcardOp(node).llgaOp());
+    }
+    GRAPH_DEBUG("  Added node ", node->kind().toQualString());
+
+    for (Value* input : node->inputs()) {
+      tensorIdToValue_.emplace(input->unique(), input);
+    }
+  }
+
+  GRAPH_DEBUG("Get Partitions");
+  std::vector<dnnl::graph::partition> partitions = g.get_partitions(policy);
+  // excluded unsupported Wildcard partitions
+  for (auto& partition : partitions) {
+    if (partition.is_supported())
+      partitions_.push_back(partition);
+  }
+
+  GRAPH_DEBUG("  Got #partitions: ", partitions_.size());
+  for (size_t partId = 0; partId < partitions_.size(); partId++) {
+    for (auto opId : partitions_[partId].get_ops()) {
+      opToOwningPartition_.add(opId, partId);
+    }
+  }
+
+  // Scanning the graph again for post processing
+  for (auto* node : graph->block()->nodes()) {
+    mayAddListConstructIntoConcatPartition(node, opToOwningPartition_);
+  }
+}
+
+bool LlgaGraphHelper::isLlgaSubgraph(const Node* node) {
+  return node->hasAttribute(attr::Subgraph) &&
+      node->kind() == prim::oneDNNFusionGroup;
+}
+
+bool LlgaGraphHelper::shouldMerge(Node* toMerge, Node* subgraph) {
+  TORCH_CHECK(
+      isLlgaSubgraph(subgraph),
+      "The consumer node does not contain a subgraph");
+  if (!shouldConsiderForMerge(toMerge)) {
+    return false;
+  }
+  return opToOwningPartition_.get(toMerge) ==
```

Comment:
`opToOwningPartition_.get(toMerge) == opToOwningPartition_.get(subgraph);` what is the purpose here ? don't we want to add new nodes to the subgraph that aren't already in the subgraph?

Reply:
`opToOwningPartition_` maps nodes^ to their respective oneDNN Graph partition ID.
We'd merge nodes if oneDNN Graph had determined that they would be in the same oneDNN Graph partition.

^ - Since PyTorch is unsupported on 32 bit platforms, it's safe to use pointers of nodes for the purpose.